crypto/src/vrf/transcript/with_rng.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::{Arc, Mutex};
+
+use randomness::{CryptoRng, RngCore};
+
+use super::{no_rng::VRFTranscript, traits::SignableTranscript};
+
+pub trait RngCoreAndCrypto: RngCore + CryptoRng {}
+
+impl<R> RngCoreAndCrypto for R where R: RngCore + CryptoRng {}
+
+/// A transcript that produces deterministic vrf signatures based on the provided rng
+#[must_use]
+#[derive(Clone)]
+pub struct VRFTranscriptWithRng<'a>(merlin::Transcript, Arc<Mutex<dyn RngCoreAndCrypto + 'a>>);
```

Comment:
I am slightly confused by having both `Arc` and a lifetime here. What is the lifetime param for?

Reply:
It's just for being as generic as possible. This is an esoteric case, in which the lifetime of the dyn object is shorter than the smart pointer holding it.